### PR TITLE
logical: Optionally filter rangefeed events to avoid replicating TTL …

### PR DESCRIPTION
**Pull request by navsetlur**
…deletes

Some of our users want the ability to avoid replicating TTL deletes from one
side of LDR to the other. This commit adds that ability to filter rangefeed
events when the bit is set on the other side. For testing, ttljob_test.go
does a lot of job scheduling wrangling to get the row TTL job to run, and
that feels unecessary for a unit test on this. This code would benefit
from a roachtest that explictly tests the TTL delete functionality with
admission control loosened

Release note (sql change): The option IGNORE_CDC_IGNORED_TTL_DELETES
can be set when creating a logical replication stream to filter out
replicating row TTL deletes. Note that the corresponding CDC flag
ttl_disable_changefeed_replicationi must be set for this to have
any effect.

Epic: CRDB-41223

### CODE REVIEW

File: docs/generated/sql/bnf/stmt_block.bnf [hunk 1]
```diff
@@ -1306,6 +1306,7 @@ unreserved_keyword ::=
 	| 'NOWAIT'
 	| 'NULLS'
 	| 'IGNORE_FOREIGN_KEYS'
```
Comment (msbutler) [inline docs/generated/sql/bnf/stmt_block.bnf:1308]: notes on commit message: while LDR is in private preview, no need to include a release note.

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go [hunk 2]
```diff
@@ -337,6 +344,12 @@ func evalLogicalReplicationOptions(
 			r.userFunctions[objName.String()] = descID
 		}
 	}
+
+	if options.FilterRangefeed == tree.DBoolTrue {
+		r.filterRangefeed = true
+	} else {
+		r.filterRangefeed = false
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:351]: nit: this else condition is unnecessary. the default value of a bool is false.

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go [hunk 3]
```diff
@@ -220,6 +222,7 @@ func createLogicalReplicationStreamPlanHook(
 				ReplicationPairs:          repPairs,
 				TableNames:                srcTableNames,
 				DefaultConflictResolution: defaultConflictResolution,
+				FilterRangefeed:           filterRangefeed,
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:225]: nit: just call options.GetFilterRangefeed() here?

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go [hunk 4]
```diff
@@ -208,6 +208,8 @@ func createLogicalReplicationStreamPlanHook(
 			defaultConflictResolution = *cr
 		}
 
+		filterRangefeed := options.GetFilterRangefeed()
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:211]: nit: I wonder if we should use a more descriptive phase throughout the  code `filterCDCIgnoredTTLDeletes`. This also makes it clear that any logic related to this option is bespoke. @dt wdyt?

File: pkg/jobs/jobspb/jobs.proto [hunk 5]
```diff
@@ -251,6 +251,8 @@ message LogicalReplicationDetails {
     int32 function_id = 2;
   }
   DefaultConflictResolution default_conflict_resolution = 7 [(gogoproto.nullable) = false];
+
+  bool FilterRangefeed = 8;
```
Comment (msbutler) [inline pkg/jobs/jobspb/jobs.proto:255]: nit: rename to `filter_rangefeed`, per go proto naming convention.

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 6]
```diff
@@ -530,4 +530,8 @@ message LogicalReplicationWriterSpec {
     // TableReplicationMetadata is a map from destination table IDs to metadata
     // containing source table descriptors and fully qualified destination table names.
     map<int32, TableReplicationMetadata> table_metadata = 8 [(gogoproto.nullable) = false];
+
+    // FilterRangefeeed is an option on whether to filter out 'OmitinRangefeed' events
+    // when processing changes in LDR
+    optional bool FilterRangefeed = 9 [(gogoproto.nullable) = false];
```
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:536]: nit: rename to filter_rangefeed, per go proto naming convention.

File: pkg/ccl/crosscluster/producer/event_stream.go [hunk 7]
```diff
@@ -160,6 +160,9 @@ func (s *eventStream) Start(ctx context.Context, txn *kv.Txn) (retErr error) {
 		// stream.
 		opts = append(opts, rangefeed.WithOriginIDsMatching(0))
 	}
+	if s.spec.WithFiltering {
```
Comment (msbutler) [inline pkg/ccl/crosscluster/producer/event_stream.go:163]: nit: this couple be simplified, without a branch: `rangefeed.WithFiltering(s.SpecWithFiltering)`

**Review by msbutler — APPROVED** [review]
In a follow up, we should add a e2e unit test for this. How does CDC test this e2e?

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go (hunk 4 above)
Comment (navsetlur) [reply pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:211]: We definitely could. I chose FilterRangefeed because under the hood, we don't have a choice in what we filter so I wanted the name to be as accurate as possible.

**Comment by navsetlur** [comment]
I don't think the TeamCity failures are due to my changes. Going to merge.

**Comment by navsetlur** [comment]
bors r=msbutler

File: pkg/sql/sem/tree/create_logical_replication.go [hunk 8]
```diff
@@ -163,6 +168,14 @@ func (o *LogicalReplicationOptions) CombineWith(other *LogicalReplicationOptions
 		}
 	}
 
+	if o.FilterRangefeed != nil {
+		if other.FilterRangefeed != nil {
+			return errors.New("IGNORE_TTL_DELETES option specified multiple times")
```
Comment (andyyang890) [inline pkg/sql/sem/tree/create_logical_replication.go:173]: drive-by comment: should this be IGNORE_CDC_IGNORED_TTL_DELETES?

Comment (navsetlur) [reply pkg/sql/sem/tree/create_logical_replication.go:173]: Good catch, will fix in another PR
